src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1802,8 +1802,8 @@ private TypeSymbol GetAccessedType(AccessExpressionSyntax syntax, IDiagnosticWri
                 syntax.IsSafeAccess || TypeValidator.ShouldWarnForPropertyMismatch(objectType),
                 diagnostics),
 
-            // TODO: We might be able use the declared type here to resolve discriminator to improve the assigned type
-            DiscriminatedObjectType => LanguageConstants.Any,
```

Comment:
I would leave these lines because tagged unions still just fall back to `any`, but we could do something like line 1806 for those as well.

Reply:
It seemed to me like the logic on line 1810 should cover this case, because `TypeValidator.AreTypesAssignable` should return true for DiscriminatedObject -> Object:

https://github.com/Azure/bicep/blob/c8be8978e71ed932ec00f176b4dd9ea8e719d685/src/Bicep.Core/TypeSystem/TypeValidator.cs#L167

Reply:
That's fair. I've put the TODO back - no harm in keeping it!